server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java

```diff
@@ -0,0 +1,193 @@
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.core.AbstractRefCounted;
+import org.elasticsearch.core.Assertions;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.ReachabilityChecker;
+import org.junit.Before;
+
+import java.util.stream.Stream;
+
+public class LeakTrackerTests extends ESTestCase {
+
+    private static final Logger logger = LogManager.getLogger();
+
+    private final TrackedObjectTestCase trackedObjectTestCase;
+    private ReachabilityChecker reachabilityChecker;
+
+    @Before
+    public void createReachabilityTracker() {
+        reachabilityChecker = new ReachabilityChecker();
+    }
+
+    public LeakTrackerTests(@Name("trackingMethod") TrackedObjectTestCase trackedObjectTestCase) {
+        this.trackedObjectTestCase = trackedObjectTestCase;
+    }
+
+    @ParametersFactory(shuffle = false)
+    public static Iterable<TrackedObjectTestCase[]> parameters() {
+        return Stream.of(
+            new PojoTrackedObjectTestCase(),
+            new ReleasableTrackedObjectTestCase(),
+            new ReferenceCountedTrackedObjectTestCase()
+        ).map(i -> new TrackedObjectTestCase[] { i }).toList();
+    }
+
+    public void testWillLogErrorWhenTrackedObjectIsNotClosed() throws Exception {
+        trackedObjectTestCase.createAndTrack(reachabilityChecker);
+        // Do not close leak before nullifying
+        trackedObjectTestCase.nullifyReference();
+        reachabilityChecker.ensureUnreachable();
+        assertBusy(() -> assertLeakDetected("LEAK: resource was not cleaned up before it was garbage-collected\\.(.*|\\s)*"));
+    }
+
+    public void testWillNotLogErrorWhenTrackedObjectIsClosed() {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        trackedObjectTestCase.createAndTrack(reachabilityChecker);
+        trackedObjectTestCase.closeLeak();
+        trackedObjectTestCase.nullifyReference();
+        // Should not detect a leak
+        reachabilityChecker.ensureUnreachable();
+    }
+
+    /**
+     * Encapsulates the lifecycle for a particular type of tracked object
+     */
+    public interface TrackedObjectTestCase {
+
+        /**
+         * Create the tracked object, implementations must
+         * - track it with the {@link LeakTracker}
+         * - register it with the passed reachability checker
+         * - retain a reference to it
+         * @param reachabilityChecker The reachability checker
+         */
+        void createAndTrack(ReachabilityChecker reachabilityChecker);
+
+        /**
+         * Nullify the retained reference
+         */
+        void nullifyReference();
+
+        /**
+         * Close the {@link LeakTracker.Leak}
+         */
+        void closeLeak();
+    }
+
+    private static class PojoTrackedObjectTestCase implements TrackedObjectTestCase {
```

Comment:
naming of static classes was dictated by the gradle testTestConventions task

Reply:
Hm I don't think we enforce this naming convention, it's just noise in IntelliJ. I would rather name these things better I think.